core/src/blocks/curl.rs

```diff
@@ -0,0 +1,239 @@
+                        .into_iter()
+                        .filter_map(|ip| match ip {
+                            std::net::IpAddr::V4(ip) => Some(ip),
+                            _ => None,
+                        })
+                        .collect::<Vec<_>>();
+                    match ipv4.len() {
+                        0 => Err(anyhow!("Could not find an ipv4 address for host: {}", d))?,
+                        _ => ipv4,
+                    }
+                }
+                Host::Ipv4(ip) => vec![ip],
+                Host::Ipv6(_) => Err(anyhow!("Ipv6 addresses are not supported."))?,
+            },
+            None => Err(anyhow!("Provided URL has an empty host"))?,
+        }
+        .into_iter()
+        .map(|ip| {
+            lazy_static! {
+                static ref RE: Regex = Regex::new(r"^(0|127|10|172|192)\..*").unwrap();
+            }
+            match RE.is_match(ip.to_string().as_str()) {
+                true => Err(anyhow!("Forbidden IP range"))?,
+                false => Ok(ip),
+            }
+        })
+        .collect::<Result<Vec<_>>>()?;
+
+        // TODO(spolu): encode query
+
+        let mut req = Request::builder().method(method).uri(url.as_str());
+
+        let headers = req.headers_mut().unwrap();
+        match headers_value {
+            Value::Object(h) => {
+                for (key, value) in h {
+                    match value {
+                        Value::String(value) => {
+                            headers.insert(
+                                header::HeaderName::from_bytes(key.as_bytes())?,
+                                header::HeaderValue::from_bytes(value.as_bytes())?,
+                            );
+                        }
+                        _ => Err(anyhow!("Header value for header {} must be a string", key))?,
+                    }
+                }
+            }
+            _ => Err(anyhow!(
+                "Returned headers must be an object with string values.",
+            ))?,
+        };
+
+        let req = match body_value {
+            Value::String(body) => req.body(Body::from(body))?,
+            Value::Null => req.body(Body::empty())?,
+            _ => Err(anyhow!("Returned body must be either a string or null."))?,
+        };
+
+        let https = HttpsConnector::new();
+        let cli = Client::builder().build::<_, hyper::Body>(https);
+
+        let res = cli.request(req).await?;
+
+        let status = res.status();
+
+        let body = hyper::body::aggregate(res).await?;
+        let mut b: Vec<u8> = vec![];
+        body.reader().read_to_end(&mut b)?;
+
+        let response_body = String::from_utf8_lossy(&b).into_owned();
+
+        let response_body = match serde_json::from_str::<serde_json::Value>(&response_body) {
+            Ok(body) => body,
+            Err(_) => Value::String(response_body),
+        };
+
+        Ok(json!({
+            "status": status.as_u16(),
+            "body": response_body,
```

Comment:
Would be nice if we did auto-conversion to JSON in the case that JSON is specified as the content type. It'd be a bummer if you keep having to add `code` blocks after `curl` to just do a `JSON.parse`.

Reply:
Oh, I may be making a bad assumption about how serde_json deals with string values in an object. I assumed they would always be of type string, but maybe it detects that the string is JSON and converts it.